Add EDOT vs. classic APM comparison doc to "Compatibility" section (#391)
elastic/opentelemetry: docs/reference/compatibility/data-streams.md

Author: I agree this would be a good addition. A follow-up issue sounds like the best way to handle it. Could you open an issue in elastic/opentelemetry to track it?

---
navigation_title: Data streams comparison
description: Learn how EDOT optimizes telemetry storage and query performance in Elastic Observability compared to classic APM.  
applies_to:
  stack:
  serverless:
    observability:
products:
  - id: cloud-serverless
  - id: observability
  - id: edot-collector
  - id: edot-sdk
---

> # EDOT data streams compared to classic APM

Reviewer: ```suggestion
# OpenTelemetry data streams compared to classic APM
```

These data streams aren't strictly tied to EDOT.

* Optimized field handling for dynamic fields (for example, `attributes`)
* Index sorting for more efficient queries   

### Metrics in TSDS

Metric data is stored using Elasticsearch’s [TSDS](docs-content://manage-data/data-store/data-streams/time-series-data-stream-tsds.md). Benefits include:

* Efficient storage using columnar compression  
* Fast aggregations 
* Automatic detection of metric dimensions (no need to manually define `time_series_dimension` in field mappings)

## Mapping optimizations with `passthrough`

The [`passthrough`](elasticsearch://reference/elasticsearch/mapping-reference/passthrough.md) field type makes EDOT attributes available at the top level, providing query compatibility with ECS-based data streams such as those used by the classic APM.

For example, while the service name is stored at `resource.attributes.service.name`, you can query it as `service.name` (the same field name used in the classic APM data stream). This allows dashboards, saved searches, and queries built for ECS to work with EDOT data without changes.

> `passthrough` also works with field aliases to map naming differences between ECS and EDOT semantic conventions. Many fields share the same name and meaning, but when names differ, aliases map them for compatibility to make migration easier. Refer to [ECS & OpenTelemetry](ecs://docs/reference/ecs-opentelemetry.md) for more information.

Reviewer 2: ```suggestion
`passthrough` also works with field aliases to map naming differences between ECS and EDOT semantic conventions. Many fields share the same name and meaning, but when names differ, aliases map them for compatibility to make migration easier. Refer to [ECS & OpenTelemetry](ecs://reference/ecs-opentelemetry.md) for more information.
```

|---|---|---|
| Index mode | General-purpose indices (logs, traces, metrics) <br><br> TSDS is not supported for classic APM | LogsDB (logs/traces), TSDS (metrics) |
| Mapping style | ECS flattened fields: nested objects are flattened into dotted field names, with all fields stored as `keyword` unless otherwise defined | Native OpenTelemetry fields with `passthrough`, preserving types and structure |
| Attribute handling | Dynamic mapping. Custom attributes stored under `labels.*` (strings) or `numeric_labels.*` (numbers); dots in field names are replaced with underscores <br><br> See [Document examples - classic APM](#classic-apm) | Dynamic mapping with native types under `attributes.*`, preserving dots in field names <br><br> See [Document examples - EDOT](#edot) |
| Query performance | Designed for general-purpose queries; performance depends on mapping complexity and data volume | Similar query performance; LogsDB and TSDS optimize for observability data patterns |
| Metric dimension definition | Manual — requires defining dimensions as fields | Automatic using TSDS — dimensions detected from document fields |

> ### Document examples

Author: @felixbarny Sure, do you have good examples for EDOT/classic APM that I could use?

---
navigation_title: Data streams comparison
description: Learn how EDOT optimizes telemetry storage and query performance in Elastic Observability compared to classic APM.  
applies_to:
  stack:
  serverless:
    observability:
products:
  - id: cloud-serverless
  - id: observability
  - id: edot-collector
  - id: edot-sdk
---

# EDOT data streams compared to classic APM

> The Elastic Distribution of OpenTelemetry (EDOT) stores telemetry data using a storage model optimized for OpenTelemetry signals. When `mapping_mode: otel` is enabled, EDOT writes logs, traces, and metrics to specialized data streams aligned with OpenTelemetry semantics to improve efficiency.

Reviewer: ```suggestion
The Elastic Distribution of OpenTelemetry (EDOT) stores telemetry data using a storage model optimized for OpenTelemetry signals. When `mapping_mode: otel` is enabled on the Elasticsearch exporter (which is the default setting), EDOT writes logs, traces, and metrics to specialized data streams aligned with OpenTelemetry semantics.
```

This architecture is designed for scalable observability workloads. It supports dynamic attributes, reduces mapping complexity, and avoids issues like mapping explosions or manual dimension setup.

EDOT uses Elasticsearch’s [Logs data stream (LogsDB)](docs-content://manage-data/data-store/data-streams/logs-data-stream.md) and [Time Series Data Streams (TSDS)](docs-content://manage-data/data-store/data-streams/time-series-data-stream-tsds.md) as storage backends. These are purpose-built to handle the scale and variety of observability data.

## Storage architecture

EDOT uses specialized index modes for each type of telemetry signal:

| Signal type | Index mode    | Storage backend |
|-------------|---------------|------------------|
| Logs        | `logsdb`      | LogsDB           |
| Traces      | `logsdb`      | LogsDB           |
| Metrics     | `time_series` | TSDS             |

> 

Reviewer: Maybe remove this section? It adds little beyond what's stated in the previous sentence.

### Logs and traces in LogsDB

> Log and trace data is stored in LogsDB, a storage engine optimized for high-ingest, semi-structured observability data. Benefits include:

Reviewer: ```suggestion
Log and trace data is stored in [LogsDB](docs-content://manage-data/data-store/data-streams/logs-data-stream.md), a storage engine optimized for high-ingest, semi-structured observability data. Benefits include:
```

> * Improved indexing performance for dynamic fields like `attributes`  

Reviewer: What do you mean by indexing performance? As in better indexing throughput? I'm not sure that's the case. In fact, LogsDB and TSDB have a slight penalty on indexing performance.

> * Field-level sorting for more efficient queries  

Reviewer: Do you mean index sorting?

Author: Yeah, I do. I think that's left form some attempt to remove some repetitions. Good catch!

> * Use of the [`passthrough`](#mapping-optimizations-with-passthrough) field type for flexible access without deep nesting  

Reviewer: I'm not sure what "flexible access without deep nesting" means. As there's a dedicated section for passthrough that applies to both LogsDB and TSDS, maybe just omit that here?

Author: I meant that using `passthrough` it can be easier to query attributes (without dealing with deeply nested structures in the mapping). I can omit it - as you pointed out, there is a dedicated section for that.

### Metrics in TSDS

> Metric data is stored using Elasticsearch’s TSDS. Benefits include:

Reviewer: ```suggestion
Metric data is stored using Elasticsearch’s [TSDS](docs-content://manage-data/data-store/data-streams/time-series-data-stream-tsds.md). Benefits include:
```

* Efficient storage using columnar compression  
* Fast aggregations 

> * Automatic detection of metric dimensions (no need to manually define field sets)  

Reviewer: ```suggestion
* Automatic detection of metric dimensions (no need to manually define `time_series_dimension` in field mappings)  
```

> ## Mapping optimizations with `passthrough`

Reviewer: Another benefit of `passthrough` fields is that they automatically set [`subobjects: false`](https://www.elastic.co/docs/reference/elasticsearch/mapping-reference/subobjects). This enables ingesting attributes and metrics that would otherwise cause mapping conflicts leading to data loss.

> ## Mapping optimizations with `passthrough`

Author: Thank you! I've rewritten the "Mapping optimizations with `passthrough`" section

OpenTelemetry attributes are stored using the [`passthrough`](elasticsearch://reference/elasticsearch/mapping-reference/passthrough.md) field type, enabling:

* Full query support  

> * Dynamic mapping without immediate mapping explosions  

Reviewer: The passthrough field type doesn't have an effect on whether or not there can be mapping explosions. The setting that helps against it is `ignore_dynamic_beyond_limit`. See also https://www.elastic.co/docs/manage-data/data-store/data-streams/logs-data-stream#logs-db-ignore-limit.

* Simplified query syntax  

> Unlike `flattened` fields, which support only keyword-style queries, `passthrough` retains field types and structure, allowing for more expressive queries.

Author: Right, better remove this sentence then.

| Factor               | Impact                                                                 |
|----------------------|------------------------------------------------------------------------|
| Index mode           | LogsDB and TSDS reduce mapping overhead and support efficient compression |
| Schema               | `mapping_mode: otel` avoids ECS-style flattening                          |
| Attribute ingestion  | `passthrough` fields support dynamic attributes with less risk of mapping explosions             |
| Sampling             | Supported natively by EDOT and Elastic APM agents                         |

Elastic compresses telemetry data on ingest. Actual disk usage depends on configuration, sampling, and data shape. [EDOT SDKs](../edot-sdks/index.md) also offer control over emitted data, letting you reduce volume by filtering spans or limiting attributes at the source.

For more details on performance and compression, refer to the [LogsDB](docs-content://manage-data/data-store/data-streams/logs-data-stream.md) and [TSDS](docs-content://manage-data/data-store/data-streams/time-series-data-stream-tsds.md) documentation.

> 

Author: Ok, sure
